cmd/gardener-scheduler/app/gardener_scheduler.go

```diff
@@ -0,0 +1,369 @@
+	if len(o.ConfigFile) == 0 {
+		return fmt.Errorf("missing GardenerScheduler config file")
+	}
+	return nil
+}
+
+// Validate validates all the required options.
+func (o *Options) validate(args []string) error {
+	if len(args) != 0 {
+		return errors.New("arguments are not supported")
+	}
+
+	return nil
+}
+
+func (o *Options) applyDefaults(in *config.SchedulerConfiguration) (*config.SchedulerConfiguration, error) {
+	external, err := scheme.Scheme.ConvertToVersion(in, schedulerconfigv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	scheme.Scheme.Default(external)
+
+	internal, err := scheme.Scheme.ConvertToVersion(external, config.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	out := internal.(*config.SchedulerConfiguration)
+
+	return out, nil
+}
+
+func (o *Options) run(ctx context.Context, cancel context.CancelFunc) error {
+	if len(o.ConfigFile) > 0 {
+		c, err := o.loadConfigFromFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		o.config = c
+	}
+
+	gardener, err := NewGardenerScheduler(o.config)
+	if err != nil {
+		return err
+	}
+
+	return gardener.Run(ctx, cancel)
+}
+
+// NewCommandStartGardenerScheduler creates a *cobra.Command object with default parameters
+func NewCommandStartGardenerScheduler(ctx context.Context, cancel context.CancelFunc) *cobra.Command {
+	opts, err := NewOptions()
+	if err != nil {
+		panic(err)
+	}
+
+	cmd := &cobra.Command{
+		Use:   "gardener-scheduler",
+		Short: "Launch the Gardener scheduler",
+		Long:  `The Gardener scheduler is a controller that tries to find the best matching seed cluster for a shoot. The scheduler takes the cloud provider and the distance between the seed (hosting the control plane) and the shoot cluster region into account.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			utilruntime.Must(opts.configFileSpecified())
+			utilruntime.Must(opts.validate(args))
+			utilruntime.Must(opts.run(ctx, cancel))
+		},
+	}
+
+	opts.config, err = opts.applyDefaults(opts.config)
+	if err != nil {
+		panic(err)
+	}
+	opts.AddFlags(cmd.Flags())
+	return cmd
+}
+
+// GardenerScheduler represents all the parameters required to start the
+// Gardener scheduler.
+type GardenerScheduler struct {
+	Config                 *config.SchedulerConfiguration
+	Identity               *gardenv1beta1.Gardener
+	GardenerNamespace      string
+	K8sGardenClient        kubernetes.Interface
+	K8sGardenInformers     gardeninformers.SharedInformerFactory
+	K8sGardenCoreInformers gardencoreinformers.SharedInformerFactory
+	KubeInformerFactory    informers.SharedInformerFactory
+	Logger                 *logrus.Logger
+	Recorder               record.EventRecorder
+	LeaderElection         *leaderelection.LeaderElectionConfig
+}
+
+// NewGardenerScheduler is the main entry point of instantiating a new Gardener Scheduler.
+func NewGardenerScheduler(cfg *config.SchedulerConfiguration) (*GardenerScheduler, error) {
+	if cfg == nil {
```

Comment:
I think rather the caller of this function should ensure that it's not `nil` - otherwise, every pointer receiving method would have to check whether its arguments are `nil` or not.